controllers/operator_controller.go

```diff
@@ -277,3 +289,21 @@ func (r *OperatorReconciler) existingBundleDeploymentUnstructured(ctx context.Co
 	}
 	return &unstructured.Unstructured{Object: unstrExistingBundleDeploymentObj}, nil
 }
+
+func (r *OperatorReconciler) verifySuccessfulBundleDeployments(ctx context.Context, name string) error {
+	bundleDeploymentExists := &rukpakv1alpha1.BundleDeployment{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: name}, bundleDeploymentExists); err != nil {
+		return fmt.Errorf("bundleDeployment %q not found %v", name, err)
+	}
+
+	conditions := bundleDeploymentExists.Status.Conditions
+	ers := []error{}
+
+	for _, c := range conditions {
+		if c.Status == metav1.ConditionFalse {
+			ers = append(ers, fmt.Errorf("error observed by Rukpak: %v", c.Message))
+		}
+	}
```

Comment:
I don't think it's safe to assume that all Conditions impact the determination of whether the Operator Ready condition is failing.

I also don't think its safe to assume that a `False` condition indicates an error (e.g. what if we add `Progressing` as a new condition type to rukpak)

I think what we should probably do here is specifically check for known condition types and implement logic based on the various possibilities.

Reply:
For now I'm checking only the `TypeHasValidBundle` and `TypeInstalled` conditions to be false explicitly. Is it sufficient, or can there ever be a situation in rukpak where the status are left `Unknown` for which we need to error?

Reply:
Would it work to check for `!= True` for now?

I think we'll want to tighten this up to handle all those possibilities though. E.g. if installed and had valid bundle are both unknown, then ready should also be unknown.

Reply:
It does, but just wanted to make sure if `Unknown` status is something for which we want to error for? Usually its an intermediate status before either failing or succeeding. But if there is a case where the status is just left Unknown, then we should be specifically checking that and erroring.

Reply:
See my other comment about thinking of this more like a mapping of BD status to Operator Ready. Not "is this BD status an error or not". Maybe that helps?

Reply:
@joelanford modified the PR based on the review comments. 

I ran a few scenarios and have some questions regarding this:

1. Previously, when we were inferring a false status as error, we were also returning the error on the controller, not just updating the status. Currently, the controller does not surface any error when bundleDeployment fails (imo it should) instead just silently only updates the status.
2. When a request is requeued for an Unknown condition, we are running the resolver again which is not necessary and could be an expensive operation by itself. Instead what we actually need, is to just wait for the rukpak to finish the steps for creating bundleDeployment. An option is to have a random backoff with requeueAfter (a fixed interval of time), but I’m not sure if that is still a valid approach to avoid running reconciliation for the same state again. 
3. Do we need to requeue when the operator’s Ready status is false if we are not triggering it with a non-nil error? I think yes, because in future there maybe a change in catalogsource’s cache to let’s say include the package we want that changes resolver's entities. But it again comes back to Q1, where the user will have no idea on why the request is getting requeued again when we do not return a valid error.

Reply:
> Currently, the controller does not surface any error when bundleDeployment fails (imo it should) instead just silently only updates the status.

I think this is actually okay. In my mind, if the operator reconcile actually fails in some way, it makes sense to return an error. But if we did everything we we're supposed to do successfully, and another component (in this case rukpak) is failing in some way, we can reflect that other component's error in our status, and then just wait to reconcile until we get another event on the the BD, the assumptions being that:
1. If we did everything successfully this time, its pretty likely we'll do everything successfully again if we reconcile immediately again with no other change in state (ergo `requeue: true` is likely to keep resulting in more `requeue: true` until some other state actually changes).
2. The other component will likely be the thing returning an error, re-reconciling, and updating the status of the underlying thing. We'll see those status updates and reconcile again based on that.

> Instead what we actually need, is to just wait for the rukpak to finish the steps for creating bundleDeployment.

Exactly. The way we wait is to requeue events on the BD (with out `Owns(BD)` watch) as rukpak progresses on it and updates its status accordingly.

> Do we need to requeue when the operator’s Ready status is false if we are not triggering it with a non-nil error? I think yes, because in future there maybe a change in catalogsource’s cache to let’s say include the package we want that changes resolver's input entities.

I think a better approach might be to have a way for the catalogsource cache to signal: "Hey, I've received a change. You should requeue _all_ of the BDs"